File: bindings/java/src/main/java/io/github/orieg/expanse/ExpanseBlobMap.java
package io.github.orieg.expanse;

import io.github.orieg.expanse.internal.ExpanseNative;
import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;
import java.lang.foreign.ValueLayout;
import java.util.Objects;
import java.util.Optional;

/**
 * High-performance off-heap map from 64-bit keys to arbitrary-length byte payloads
 * backed by polymorphic 64-bit value slots and chunked slab arenas.
 */
public final class ExpanseBlobMap implements AutoCloseable {

    /**
     * A retrieved blob record containing byte payload, 32-bit hot metadata, and storage tag.
     *
     * @param data payload byte array
     * @param hotMeta 32-bit hot metadata word
     * @param isInline true if payload was stored inline in the value slot (<= 7 bytes)
     */
    public record BlobRecord(byte[] data, int hotMeta, boolean isInline) {}

    private MemorySegment handle;
    private boolean closed = false;

    /**
     * Creates a new empty {@link ExpanseBlobMap} with default 2 MiB slab chunks.
     */
    public ExpanseBlobMap() {
        this(0);
    }

    /**
     * Creates a new empty {@link ExpanseBlobMap} with custom slab chunk capacity in bytes.
     *
     * @param chunkSize slab chunk size in bytes
     */
    public ExpanseBlobMap(long chunkSize) {
        try {
            this.handle = (MemorySegment) ExpanseNative.MH_expanse_blob_map_new.invokeExact(chunkSize);
            if (handle.equals(MemorySegment.NULL)) {
                throw new OutOfMemoryError("Failed to allocate native ExpanseBlobMap");
            }
        } catch (Throwable t) {
            throw new RuntimeException("Failed creating ExpanseBlobMap", t);
        }
    }

    private void checkOpen() {
        if (closed || handle.equals(MemorySegment.NULL)) {
            throw new IllegalStateException("ExpanseBlobMap has been closed");
        }
    }

    /**
     * Inserts a key-blob mapping with default hot metadata (0).
     *
     * @param key 64-bit unsigned key
     * @param data payload byte array
     * @return true on success
     */
    public boolean insert(long key, byte[] data) {
        return insert(key, data, 0);
    }

    /**
     * Inserts a key-blob mapping with 32-bit hot metadata.
     *
     * @param key 64-bit unsigned key
     * @param data payload byte array
     * @param hotMeta 32-bit hot metadata
     * @return true on success
     */
    public boolean insert(long key, byte[] data, int hotMeta) {
        Objects.requireNonNull(data, "data must not be null");
        checkOpen();
        if (data.length == 0) {
            return insert(key, MemorySegment.NULL, 0, hotMeta);
        }
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment seg = arena.allocate(ValueLayout.JAVA_BYTE, data.length);
            MemorySegment.copy(data, 0, seg, ValueLayout.JAVA_BYTE, 0, data.length);
            return insert(key, seg, data.length, hotMeta);
        }
    }

    /**
     * Inserts a key-blob mapping from raw native memory segment.
     *
     * @param key 64-bit key
     * @param dataSegment memory segment pointing to payload bytes
     * @param len byte length
     * @param hotMeta 32-bit hot metadata
     * @return true on success
     */
    public boolean insert(long key, MemorySegment dataSegment, long len, int hotMeta) {
        checkOpen();
        try {
            return (boolean) ExpanseNative.MH_expanse_blob_map_insert.invokeExact(
                    handle, key, dataSegment, len, hotMeta);
        } catch (Throwable t) {
            throw new RuntimeException(t);
        }
    }

    /**
     * Retrieves the blob record for the given key.
     *
     * @param key 64-bit key
     * @return Optional containing BlobRecord if key is present
     */
    public Optional<BlobRecord> get(long key) {
        checkOpen();
        try (Arena arena = Arena.ofConfined()) {
            // Allocate 24-byte struct ExpanseBlobView: ptr(8), len(8), hot_meta(4), is_inline(1), pad(3)
            MemorySegment outView = arena.allocate(24, 8);
            boolean found = (boolean) ExpanseNative.MH_expanse_blob_map_get.invokeExact(handle, key, outView);
            if (!found) {
                return Optional.empty();
            }

            MemorySegment ptr = outView.get(ValueLayout.ADDRESS, 0);
            long len = outView.get(ValueLayout.JAVA_LONG, 8);
            int hotMeta = outView.get(ValueLayout.JAVA_INT, 16);
            boolean isInline = outView.get(ValueLayout.JAVA_BOOLEAN, 20);

            byte[] bytes;
            if (len == 0 || ptr.equals(MemorySegment.NULL)) {
                bytes = new byte[0];
            } else {
                bytes = new byte[(int) len];
                MemorySegment dataSeg = ptr.reinterpret(len);
                MemorySegment.copy(dataSeg, ValueLayout.JAVA_BYTE, 0, bytes, 0, (int) len);
            }

            return Optional.of(new BlobRecord(bytes, hotMeta, isInline));
        } catch (Throwable t) {
            throw new RuntimeException(t);
        }
    }

    /**
     * Retrieves only the payload bytes for the given key.
     *
     * @param key 64-bit key
     * @return byte array if present, null otherwise
     */
    public byte[] getBytes(long key) {
        return get(key).map(BlobRecord::data).orElse(null);
    }

    /**
     * Removes a key from the map.
     *
     * @param key 64-bit key
     * @return true if key was present and removed
     */
    public boolean remove(long key) {
        checkOpen();
        try {
            return (boolean) ExpanseNative.MH_expanse_blob_map_remove.invokeExact(handle, key);
        } catch (Throwable t) {
            throw new RuntimeException(t);
        }
    }

    /**
     * Checks if key exists in the map.
     *
     * @param key 64-bit key
     * @return true if key is present
     */
    public boolean containsKey(long key) {
        checkOpen();
        try {
            return (boolean) ExpanseNative.MH_expanse_blob_map_contains_key.invokeExact(handle, key);
        } catch (Throwable t) {
            throw new RuntimeException(t);
        }
    }

    /**
     * Returns the number of entries in the map.
     *
     * @return count of stored keys
     */
    public long len() {
        if (closed || handle.equals(MemorySegment.NULL)) {
            return 0;
        }
        try {
            return (long) ExpanseNative.MH_expanse_blob_map_len.invokeExact(handle);
        } catch (Throwable t) {
            throw new RuntimeException(t);
        }
    }

    /**
     * Returns true if map is empty.
     *
     * @return true if empty
     */
    public boolean isEmpty() {
        return len() == 0;
    }

    /**
     * Returns total off-heap bytes used by index and slab arena.
     *
     * @return heap memory in bytes
     */
    public long memUsed() {
        if (closed || handle.equals(MemorySegment.NULL)) {
            return 0;
        }
        try {
            return (long) ExpanseNative.MH_expanse_blob_map_mem_used.invokeExact(handle);
        } catch (Throwable t) {
            throw new RuntimeException(t);
        }
    }

    /**
     * Runs in-place arena garbage collection and compaction.
     *
     * @return true if compaction succeeded
     */
    public boolean compact() {
        checkOpen();
        try {
            return (boolean) ExpanseNative.MH_expanse_blob_map_compact.invokeExact(handle);
        } catch (Throwable t) {
            throw new RuntimeException(t);
        }
    }

    /**
     * Removes all entries and resets the slab arena.
     */
    public void clear() {
        if (closed || handle.equals(MemorySegment.NULL)) {
            return;
        }
        try {
            ExpanseNative.MH_expanse_blob_map_clear.invokeExact(handle);
        } catch (Throwable t) {
            throw new RuntimeException(t);
        }
    }

    @Override
    public void close() {
        if (!closed && !handle.equals(MemorySegment.NULL)) {
            try {
                ExpanseNative.MH_expanse_blob_map_free.invokeExact(handle);
            } catch (Throwable t) {
                throw new RuntimeException(t);
            } finally {
                handle = MemorySegment.NULL;
                closed = true;
            }
        }
    }
}

File: bindings/dotnet/src/Expanse.NET/ExpanseBlobMap.cs
using System;
using System.Collections.Generic;
using System.Runtime.InteropServices;
using Expanse.Native;

namespace Expanse;

/// <summary>
/// High-performance off-heap map from 64-bit integer keys to arbitrary-length byte payloads
/// backed by inline polymorphic 64-bit value slots and chunked slab arenas.
/// </summary>
public sealed class ExpanseBlobMap : IDisposable
{
    private SafeExpanseBlobMapHandle _handle;
    private bool _disposed;

    /// <summary>
    /// Creates a new empty <see cref="ExpanseBlobMap"/> with default 2 MiB slab chunks.
    /// </summary>
    public ExpanseBlobMap() : this(0) { }

    /// <summary>
    /// Creates a new empty <see cref="ExpanseBlobMap"/> with custom slab chunk capacity in bytes.
    /// </summary>
    /// <param name="chunkSize">Chunk size in bytes (0 for default 2 MiB).</param>
    public ExpanseBlobMap(nuint chunkSize)
    {
        _handle = NativeMethods.expanse_blob_map_new(chunkSize);
        if (_handle.IsInvalid)
        {
            throw new OutOfMemoryError("Failed to allocate native ExpanseBlobMap");
        }
    }

    /// <summary>
    /// Gets the underlying native <see cref="SafeExpanseBlobMapHandle"/>.
    /// </summary>
    public SafeExpanseBlobMapHandle Handle
    {
        get
        {
            ThrowIfDisposed();
            return _handle;
        }
    }

    private void ThrowIfDisposed()
    {
        ObjectDisposedException.ThrowIf(_disposed || _handle.IsInvalid || _handle.IsClosed, this);
    }

    /// <summary>
    /// Inserts or updates a key-blob mapping with optional 32-bit hot metadata.
    /// </summary>
    /// <param name="key">The 64-bit unsigned key.</param>
    /// <param name="payload">The byte payload to store.</param>
    /// <param name="hotMeta">32-bit hot metadata stored directly in index.</param>
    public unsafe void Set(ulong key, ReadOnlySpan<byte> payload, uint hotMeta = 0)
    {
        ThrowIfDisposed();
        if (payload.Length == 0)
        {
            NativeMethods.expanse_blob_map_insert(_handle, key, null, 0, hotMeta);
            return;
        }
        fixed (byte* pData = payload)
        {
            NativeMethods.expanse_blob_map_insert(_handle, key, pData, (nuint)payload.Length, hotMeta);
        }
    }

    /// <summary>
    /// Inserts or updates a key-blob mapping with byte array and optional hot metadata.
    /// </summary>
    public void Set(ulong key, byte[] payload, uint hotMeta = 0)
    {
        ArgumentNullException.ThrowIfNull(payload);
        Set(key, payload.AsSpan(), hotMeta);
    }

    /// <summary>
    /// Inserts or updates a key-blob mapping, returning <c>true</c> on success.
    /// </summary>
    public unsafe bool Insert(ulong key, ReadOnlySpan<byte> payload, uint hotMeta = 0)
    {
        ThrowIfDisposed();
        if (payload.Length == 0)
        {
            return NativeMethods.expanse_blob_map_insert(_handle, key, null, 0, hotMeta);
        }
        fixed (byte* pData = payload)
        {
            return NativeMethods.expanse_blob_map_insert(_handle, key, pData, (nuint)payload.Length, hotMeta);
        }
    }

    /// <summary>
    /// Attempts to retrieve the zero-copy blob payload and hot metadata for the given key.
    /// </summary>
    /// <param name="key">The 64-bit key.</param>
    /// <param name="payload">Zero-copy span pointing directly to the payload.</param>
    /// <param name="hotMeta">The retrieved 32-bit hot metadata.</param>
    /// <returns><c>true</c> if the key is present; otherwise <c>false</c>.</returns>
    public unsafe bool TryGet(ulong key, out ReadOnlySpan<byte> payload, out uint hotMeta)
    {
        ThrowIfDisposed();
        if (NativeMethods.expanse_blob_map_get(_handle, key, out NativeMethods.NativeBlobView view))
        {
            hotMeta = view.HotMeta;
            payload = view.Len == 0 || view.Ptr == IntPtr.Zero
                ? ReadOnlySpan<byte>.Empty
                : new ReadOnlySpan<byte>((void*)view.Ptr, (int)view.Len);
            return true;
        }
        hotMeta = 0;
        payload = default;
        return false;
    }

    /// <summary>
    /// Attempts to retrieve the zero-copy blob payload for the given key.
    /// </summary>
    public bool TryGet(ulong key, out ReadOnlySpan<byte> payload) => TryGet(key, out payload, out _);

    /// <summary>
    /// Attempts to retrieve the blob payload as a managed byte array.
    /// </summary>
    public bool TryGetBytes(ulong key, out byte[] payload, out uint hotMeta)
    {
        if (TryGet(key, out ReadOnlySpan<byte> span, out hotMeta))
        {
            payload = span.ToArray();
            return true;
        }
        payload = Array.Empty<byte>();
        return false;
    }

    /// <summary>
    /// Attempts to retrieve the blob payload as a managed byte array.
    /// </summary>
    public bool TryGetBytes(ulong key, out byte[] payload) => TryGetBytes(key, out payload, out _);

    /// <summary>
    /// Retrieves the payload byte array for the key, or <c>null</c> if absent.
    /// </summary>
    public byte[]? GetBytes(ulong key)
    {
        return TryGetBytes(key, out byte[] bytes) ? bytes : null;
    }

    /// <summary>
    /// Removes a key from the map.
    /// </summary>
    /// <param name="key">The 64-bit key.</param>
    /// <returns><c>true</c> if the key was present and removed; otherwise <c>false</c>.</returns>
    public bool Remove(ulong key)
    {
        ThrowIfDisposed();
        return NativeMethods.expanse_blob_map_remove(_handle, key);
    }

    /// <summary>
    /// Checks whether the map contains the specified key.
    /// </summary>
    /// <param name="key">The 64-bit key.</param>
    /// <returns><c>true</c> if present; otherwise <c>false</c>.</returns>
    public bool Contains(ulong key) => ContainsKey(key);

    /// <summary>
    /// Checks whether the map contains the specified key.
    /// </summary>
    public bool ContainsKey(ulong key)
    {
        ThrowIfDisposed();
        return NativeMethods.expanse_blob_map_contains_key(_handle, key);
    }

    /// <summary>
    /// Gets the number of entries stored in the map (capped at <see cref="int.MaxValue"/>).
    /// </summary>
    public int Count
    {
        get
        {
            ulong len = LongCount;
            return len > int.MaxValue ? int.MaxValue : (int)len;
        }
    }

    /// <summary>
    /// Gets the exact 64-bit count of entries stored in the map.
    /// </summary>
    public ulong LongCount
    {
        get
        {
            ThrowIfDisposed();
            return NativeMethods.expanse_blob_map_len(_handle);
        }
    }

    /// <summary>
    /// Gets whether the map is empty.
    /// </summary>
    public bool IsEmpty => LongCount == 0;

    /// <summary>
    /// Gets the total off-heap heap bytes used by the index and slab arena.
    /// </summary>
    public nuint MemoryUsed
    {
        get
        {
            ThrowIfDisposed();
            return NativeMethods.expanse_blob_map_mem_used(_handle);
        }
    }

    /// <summary>
    /// Removes all entries and resets the slab arena.
    /// </summary>
    public void Clear()
    {
        ThrowIfDisposed();
        NativeMethods.expanse_blob_map_clear(_handle);
    }

    /// <summary>
    /// Runs in-place garbage collection and compaction, consolidating live payloads and freeing dead chunks.
    /// </summary>
    /// <returns><c>true</c> if compaction succeeded.</returns>
    public bool Compact()
    {
        ThrowIfDisposed();
        return NativeMethods.expanse_blob_map_compact(_handle);
    }

    /// <summary>
    /// Prunes entries that match the given predicate, removes them, and triggers compaction.
    /// </summary>
    /// <param name="predicate">A delegate taking (key, hotMeta) returning <c>true</c> if the entry should be pruned.</param>
    /// <returns>The number of entries pruned.</returns>
    public ulong Prune(Func<ulong, uint, bool> predicate)
    {
        ArgumentNullException.ThrowIfNull(predicate);
        ThrowIfDisposed();

        var toPrune = new List<ulong>();

        NativeMethods.ExpanseScanCallback scanCb = (ulong key, NativeMethods.NativeBlobView view, IntPtr userCtx) =>
        {
            if (predicate(key, view.HotMeta))
            {
                toPrune.Add(key);
            }
            return true;
        };

        NativeMethods.expanse_blob_map_scan_filtered(
            _handle,
            0,
            ulong.MaxValue,
            null,
            scanCb,
            IntPtr.Zero);

        ulong prunedCount = 0;
        foreach (ulong key in toPrune)
        {
            if (NativeMethods.expanse_blob_map_remove(_handle, key))
            {
                prunedCount++;
            }
        }

        if (prunedCount > 0)
        {
            NativeMethods.expanse_blob_map_compact(_handle);
        }

        return prunedCount;
    }

    /// <summary>
    /// Executes a range scan over keys in [<paramref name="startKey"/>, <paramref name="endKey"/>]
    /// with optional hot metadata predicate filtering.
    /// </summary>
    public unsafe ulong ScanFiltered(
        ulong startKey,
        ulong endKey,
        Func<ulong, uint, bool>? predicate,
        Action<ulong, ReadOnlySpan<byte>, uint> callback)
    {
        ArgumentNullException.ThrowIfNull(callback);
        ThrowIfDisposed();

        NativeMethods.ExpansePredicateCallback? predCb = predicate != null
            ? (ulong key, uint meta, IntPtr ctx) => predicate(key, meta)
            : null;

        NativeMethods.ExpanseScanCallback scanCb = (ulong key, NativeMethods.NativeBlobView view, IntPtr ctx) =>
        {
            ReadOnlySpan<byte> span = view.Len == 0 || view.Ptr == IntPtr.Zero
                ? ReadOnlySpan<byte>.Empty
                : new ReadOnlySpan<byte>((void*)view.Ptr, (int)view.Len);
            callback(key, span, view.HotMeta);
            return true;
        };

        nuint count = NativeMethods.expanse_blob_map_scan_filtered(
            _handle,
            startKey,
            endKey,
            predCb,
            scanCb,
            IntPtr.Zero);

        return (ulong)count;
    }

    /// <summary>
    /// Frees the unmanaged memory allocated by this blob map.
    /// </summary>
    public void Dispose()
    {
        if (!_disposed)
        {
            _handle.Dispose();
            _disposed = true;
        }
    }
}
